### PR TITLE
[ft-template] Add cookbook for modifying hyperparameters

### PR DESCRIPTION
**Pull request by SumanthRH**
# What does this PR do?

This PR adds a simple guide to modify different hyperparameters while using the finetuning template.

Closes https://anyscale1.atlassian.net/browse/AE-846

### CODE REVIEW

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    "\n",
+    "```yaml\n",
+    "learning_rate: 1e-4\n",
+    "lr_scheduler_type: cosine\n",
+    "num_warmup_steps: 10\n",
+    "```\n",
+    "\n",
+    "In the above config, the training run would use a cosine learning rate schedule (the default) with an initial warmup of 10 steps (the default). The peak learning rate would be 1e-4 (the value specified). \n",
+    "\n",
+    "We support both `'linear'` and `'cosine'` schedules. "
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb:57]: in here take a screen shot of a cosine learning rate schedule so it quickly clicks to the user.

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    "## Batch size\n",
+    "\n",
+    "The batch size for training and validation depends on the below parameters:\n",
+    "\n",
+    "```yaml\n",
+    "num_devices: 8\n",
+    "train_batch_size_per_device: 16\n",
+    "eval_batch_size_per_device: 16\n",
+    "```\n",
+    "The effective batch size for training would be `train_batch_size_per_device * num_devices`. For the hardware you specify, the amount you can push `train_batch_size_per_device` / `eval_batch_size_per_device` depends on dataset statistics (average sequence length) and the context length used. For a context length of 512 and the default NVIDIA A10 GPUs, the per-device batch size of 16 is a good default. "
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb:73]:
```suggestion
    "The effective batch size for training would be `train_batch_size_per_device * num_devices`. For the hardware you specify, the amount you can push `train_batch_size_per_device` or `eval_batch_size_per_device` depends on dataset statistics (average sequence length) and the context length used. For a context length of 512 and the default NVIDIA A10 GPUs, the per-device batch size of 16 is a good default. "
```

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb [hunk 3]
```diff
@@ -0,0 +1,119 @@
+{
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb:1]: Also include num_epochs, talk about validation at each epoch, talk about controlling the checkpointing every epoch and num_keep, etc.

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    - q_proj
+    - v_proj
+    - k_proj
+    - o_proj
+    - gate_proj
+    - up_proj
+    - down_proj
+    - embed_tokens
+    - lm_head
+  task_type: "CAUSAL_LM"
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:65]: In here tell the user that by default we recommend to train on all target modules for the best performance.

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+learning_rate: 1e-4
+lr_scheduler_type: cosine
+num_warmup_steps: 10
+```
+
+In the above config, the training run would use a cosine learning rate schedule (the default) with an initial warmup of 10 steps (the default). The peak learning rate would be 1e-4 (the value specified). 
+
+We support both `'linear'` and `'cosine'` schedules. The learning rate schedules have been plotted below with `num_warmup_steps=10` and for 1000 training steps.
+
+<p align="center">
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:36]: The image doesn't get rendered here because the path makes sense once this PR is merged into `main`.

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb
[the remark this answers is not in the document]
Comment (SumanthRH) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.ipynb:1]: Done!

File: templates/fine-tune-llm_v2/assets/schedulers.png
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/assets/schedulers.png:1]: you put this file under `cookbooks/foo/assets` (look at the existing cookbook for example)

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+<img src="https://raw.githubusercontent.com/anyscale/templates/main/templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/../../assets/schedulers.png" alt="cosine" height="300"/>
+</p>
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:38]: Use relative paths: Look at the other cookbook for example.

**Review by kouroshHakha — COMMENTED** [review]
a few more nits:

File: templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md (hunk 6 above)
Comment (SumanthRH) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:38]: Hmm already have actually (you can see it in the .ipynb). automatic doc generation messed it up. Let me see what I can do.

Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:38]: wait, let me give you pointers

Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:38]: https://github.com/anyscale/templates/blob/main/templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb

Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md:38]: put the assets directory next to the cookbook and I think it will work

Comment (SumanthRH) [reply templates/fine-tune-llm_v2/cookbooks/modifying_hyperparameters/README.md]: Done! Image path from current directory works fine but relative path outside gets resolved differently :/

File: templates/fine-tune-llm_v2/assets/schedulers.png
Comment (SumanthRH) [reply templates/fine-tune-llm_v2/assets/schedulers.png:1]: Better! Done!

**Review by kouroshHakha — APPROVED** [review]
